bindings/c/test/unit/unit_tests.cpp

```diff
@@ -0,0 +1,1787 @@
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+    fdb_check(f2.set_callback(
+        +[](FDBFuture *f, void *param) {
+          fdb_error_t err = fdb_future_get_error(f);
+          if (err != 1101) { // operation_cancelled
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+    fdb_check(f3.set_callback(
+        +[](FDBFuture *f, void *param) {
+          fdb_error_t err = fdb_future_get_error(f);
+          if (err != 1101) { // operation_cancelled
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+    fdb_check(f4.set_callback(
+        +[](FDBFuture *f, void *param) {
+          fdb_error_t err = fdb_future_get_error(f);
+          if (err != 1101) { // operation_cancelled
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+
+    event->wait();
+    break;
+  }
+
+  // Reset available number of watches.
+  max_watches = 10000;
+  fdb_check(fdb_database_set_option(db, FDB_DB_OPTION_MAX_WATCHES,
+                                    (const uint8_t *)&max_watches, 8));
+}
+
+TEST_CASE("fdb_transaction_watch") {
+  fdb::Transaction tr(db);
+
+  struct Context {
+    FdbEvent event;
+  };
+  Context context;
+
+  while (1) {
+    fdb::EmptyFuture f1 = tr.watch(KEY("foo"));
+    fdb::EmptyFuture f2 = tr.commit();
+
+    fdb_error_t err = wait_future(f2);
+    if (err) {
+      fdb::EmptyFuture f3 = tr.on_error(err);
+      fdb_check(wait_future(f3));
+      continue;
+    }
+
+    fdb_check(f1.set_callback(
+        +[](FDBFuture *, void *param) {
+          auto *context = static_cast<Context *>(param);
+          context->event.set();
+        },
+        &context));
+    break;
+  }
+
+  // Insert data for key "foo" to trigger the watch.
+  insert_data(db, create_data({ { "foo", "bar" } }));
+  context.event.wait();
+}
+
+TEST_CASE("fdb_transaction_cancel") {
+  // Cannot use transaction after cancelling it...
+  fdb::Transaction tr(db);
+  tr.cancel();
+  fdb::ValueFuture f1 = tr.get("foo", /* snapshot */ false);
+  CHECK(wait_future(f1) == 1025); // transaction_cancelled
+
+  // ... until the transaction has been reset.
+  tr.reset();
+  fdb::ValueFuture f2 = tr.get("foo", /* snapshot */ false);
+  fdb_check(wait_future(f2));
+}
+
+TEST_CASE("fdb_error_predicate") {
```

Comment:
@sfc-gh-anoyes you may want to review these, feel free to remove lines or suggest additions. I just went based on what I thought made sense more-or-less.

Reply:
I think this looks good